File: 74751_lab5_task3.py
class Device:
    def __init__(self, name, power_consumption):
        self.name = name
        self.power_consumption = power_consumption

class HospitalCircuit:
    def __init__(self, circuit_id="Unknown", voltage_rating=220.0, obj=None):
        if obj is not None and isinstance(obj, HospitalCircuit):
            self.circuit_id = obj.circuit_id + "_copy"
            self.voltage_rating = obj.voltage_rating
            self.current_load = obj.current_load
            self.connected_devices = [Device(d.name, d.power_consumption) for d in obj.connected_devices]
        else:
            self.circuit_id = circuit_id
            self.voltage_rating = voltage_rating
            self.current_load = 0.0
            self.connected_devices = []

    def __del__(self):
        print(f"Circuit {self.circuit_id} is taken offline.")

    def add_device(self, device_name, power):
        self.connected_devices.append(Device(device_name, power))
        self.current_load += power / self.voltage_rating
        print(f"Device {device_name} added.")

    def remove_device(self, device_name):
        for d in self.connected_devices:
            if d.name == device_name:
                self.current_load -= d.power_consumption / self.voltage_rating
                self.connected_devices.remove(d)
                print(f"Device {device_name} removed.")
                return
        print("Device not found.")

    def calculate_total_power_consumption(self):
        return sum(d.power_consumption for d in self.connected_devices)

    def check_for_overload(self, max_current_load):
        return self.current_load > max_current_load

    def display_circuit_information(self):
        print("--- Circuit Information ---")
        print(f"ID: {self.circuit_id} | Voltage: {self.voltage_rating} V")
        print(f"Current Load: {self.current_load:.2f} A")
        print(f"Connected Devices: {len(self.connected_devices)}")
        for d in self.connected_devices:
            print(f"  - {d.name} ({d.power_consumption} W)")

if __name__ == "__main__":
    c1 = HospitalCircuit("ICU-Wiring-01", 220.0)
    c1.add_device("Ventilator", 500.0)
    c1.add_device("Monitor", 100.0)
    
    c2 = HospitalCircuit(obj=c1)
    
    c1.display_circuit_information()
    print(f"Total Power Consumption: {c1.calculate_total_power_consumption()} W")
    
    if c1.check_for_overload(10.0):
        print("WARNING: Overload condition detected!")
    else:
        print("Circuit load is within safe limits.")
        
    c1.remove_device("Monitor")

File: 74751_lab5_task3.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Device {
    string name;
    double powerConsumption; // in Watts
};

class HospitalCircuit {
private:
    string circuitID;
    double voltageRating;
    double currentLoad; // in Amps
    vector<Device> connectedDevices;

public:
    // Default constructor
    HospitalCircuit() : circuitID("Unknown"), voltageRating(220.0), currentLoad(0.0) {}

    // Parameterized constructor
    HospitalCircuit(string id, double voltage) 
        : circuitID(id), voltageRating(voltage), currentLoad(0.0) {}

    // Copy constructor
    HospitalCircuit(const HospitalCircuit& other) 
        : circuitID(other.circuitID + "_copy"), voltageRating(other.voltageRating),
          currentLoad(other.currentLoad), connectedDevices(other.connectedDevices) {}

    // Destructor
    ~HospitalCircuit() {
        cout << "Circuit " << circuitID << " is taken offline." << endl;
    }

    void addDevice(string deviceName, double power) {
        connectedDevices.push_back({deviceName, power});
        currentLoad += power / voltageRating;
        cout << "Device " << deviceName << " added." << endl;
    }

    void removeDevice(string deviceName) {
        for (auto it = connectedDevices.begin(); it != connectedDevices.end(); ++it) {
            if (it->name == deviceName) {
                currentLoad -= it->powerConsumption / voltageRating;
                cout << "Device " << deviceName << " removed." << endl;
                connectedDevices.erase(it);
                return;
            }
        }
        cout << "Device not found." << endl;
    }

    double calculateTotalPowerConsumption() const {
        double totalPower = 0.0;
        for (const auto& device : connectedDevices) {
            totalPower += device.powerConsumption;
        }
        return totalPower;
    }

    bool checkForOverload(double maxCurrentLoad) const {
        return currentLoad > maxCurrentLoad;
    }

    void displayCircuitInformation() const {
        cout << "--- Circuit Information ---" << endl;
        cout << "ID: " << circuitID << " | Voltage: " << voltageRating << " V" << endl;
        cout << "Current Load: " << currentLoad << " A" << endl;
        cout << "Connected Devices: " << connectedDevices.size() << endl;
        for (const auto& device : connectedDevices) {
            cout << "  - " << device.name << " (" << device.powerConsumption << " W)" << endl;
        }
    }
};

int main() {
    HospitalCircuit c1("ICU-Wiring-01", 220.0);
    c1.addDevice("Ventilator", 500.0);
    c1.addDevice("Monitor", 100.0);
    
    HospitalCircuit c2 = c1;

    c1.displayCircuitInformation();
    cout << "Total Power Consumption: " << c1.calculateTotalPowerConsumption() << " W" << endl;
    if (c1.checkForOverload(10.0)) {
        cout << "WARNING: Overload condition detected!" << endl;
    } else {
        cout << "Circuit load is within safe limits." << endl;
    }

    c1.removeDevice("Monitor");
    
    return 0;
}
